userspace/falco/webserver.cpp

```diff
@@ -68,6 +69,28 @@ void falco_webserver::start(
             res.set_content(versions_json_str, "application/json");
         });
 
+    if (!metrics_collectors.empty())
+    {
+        libs::metrics::prometheus_metrics_converter prometheus_metrics_converter;
+
+        m_server->Get("/metrics",
+            [metrics_collectors, prometheus_metrics_converter](const httplib::Request &, httplib::Response &res) {
+                std::string prometheus_text;
+
+                for (auto metrics_collector: metrics_collectors) {
+                    metrics_collector.snapshot();
+                    auto metrics_snapshot = metrics_collector.get_metrics();
+
+                    for (auto& metric: metrics_snapshot)
```

Comment:
Could we also add all the wrapper metrics already in, see for example https://github.com/falcosecurity/falco/issues/3131#issuecomment-1984818730 (see the libs unit tests) or the website https://falco.org/docs/metrics/falco-metrics/ or run Falco w/ the output rule metrics.

Or would you just want to do that in another PR? Also ok for me.

Some more notes:

For Prometheus we do not need the `_prev` metrics values and even the drop percentage is not needed as it can be calculated via Prometheus queries, hence not a 1:1 translation in terms of the wrapper or additional metrics fields.

In addition, we distinguish between `falco.` and `scap.` prefixes -- probably best to preserve that and as namespace I personally would prefer `falcosecurity`.

Reply:
From a code organization point of view we can also iterate as it would be nice to not add metrics logic into the webserver file itself and rather maybe create a more formal `falco_metrics_collector` class or something else.

Or maybe you planned that already in a follow up PR.

Reply:
Additional note I forgot: It appears we need to move the code logic to the `stats_writer::collector::collector` so that the metrics get refreshed and reposted to `/metrics` every x interval.

Reply:
> In addition, we distinguish between falco. and scap. prefixes -- probably best to preserve that and as namespace I personally would prefer falcosecurity.

Agree with `falcosecurity`.

From what i see, we are refreshing the stats when the `/metrics` endpoint is called, instead of passing through the `stats_writer::collector::collector`.
This ensures that we follow Prometheus settings (ie: if one setups prometheus to scrape every 2s, we honor that even if stats collector interval is higher).
Don't know if this is the behavior we want though.

Reply:
I can see the dilemma. At the same time I also don't see a problem with just recommending to having the same Falco metrics interval as the Prometheus scrape interval? How is this solved in other projects?  At the end of the day metrics time series are always discrete ... I would also be fine to refresh the metrics in the Prometheus case whenever `/metrics` was called as @FedeDP suggested.

Reply:
One thing that comes to mind, doesn't the stats_writer class contain the required metrics generation logic ?

Reply:
The metrics wrapper fields (the ones not part of the `libs_metrics_collector`) are easy to duplicate if needed, or we create some sort of more formal class `falco-metrics_collector` ...

Reply:
> At the same time I also don't see a problem with just recommending to having the same Falco metrics interval as the Prometheus scrape interval?

Yep, but:
* `/metrics` endpoint is the only "interactive" metrics generator we have, since `output_rule` and `output_file` are passive. Therefore, it makes sense that is up to the caller to decide the frequency.
* also, consider the case where prometheus is setup to scrape every 5seconds; consider also that `output_file` is enabled too. This has a cost in terms of IO perf if user has to set 5s `metrics.interval` in the config file

At the same time, i quite don't like having a metric that does not obey to `metrics.interval` though...

I would love to gather more feedback! cc @falcosecurity/falco-maintainers

Reply:
Maybe we need to just make everything configurable, for example, we could add sub metrics configs similar to below:

```
metrics:
  enabled: false
  interval: 1h
  # Typically, in production, you only use `output_rule` or `output_file`, but not both. 
  # However, if you have a very unique use case, you can use both together.
  output_rule: true
  # output_file: /tmp/falco_stats.jsonl
  output_prometheus:
    # Will only be enabled when `metrics.enabled` is set to true
    enabled: false
    # Take a new metrics snapshot automatically after each Prometheus server scrape
    auto_refresh_enabled: false
    # Alternative to `auto_refresh_enabled` refreshes /metrics with new metrics snapshot every x interval; supersedes and is separate from the global `metrics.interval` setting
    refresh_interval: 5m
  resource_utilization_enabled: true
  state_counters_enabled: true
  kernel_event_counters_enabled: true
  libbpf_stats_enabled: true
  convert_memory_to_mb: true
  include_empty_values: false
```